File: backend/app/domain/alert_strategies.py
from dataclasses import dataclass


@dataclass(frozen=True)
class AlertDecision:
    severity: str
    channel: str
    responder: str


class AlertStrategy:
    def evaluate(self, signal: dict) -> AlertDecision:
        return AlertDecision("P3", "dashboard", "sre-oncall")


class RdbmsAlertStrategy(AlertStrategy):
    def evaluate(self, signal: dict) -> AlertDecision:
        return AlertDecision("P0", "pager", "database-oncall")


class McpAlertStrategy(AlertStrategy):
    def evaluate(self, signal: dict) -> AlertDecision:
        return AlertDecision("P1", "pager", "platform-oncall")


class CacheAlertStrategy(AlertStrategy):
    def evaluate(self, signal: dict) -> AlertDecision:
        return AlertDecision("P2", "slack", "cache-oncall")


class QueueAlertStrategy(AlertStrategy):
    def evaluate(self, signal: dict) -> AlertDecision:
        return AlertDecision("P1", "pager", "async-oncall")


STRATEGIES: dict[str, AlertStrategy] = {
    "RDBMS": RdbmsAlertStrategy(),
    "MCP_HOST": McpAlertStrategy(),
    "CACHE": CacheAlertStrategy(),
    "QUEUE": QueueAlertStrategy(),
}


def get_alert_strategy(component_type: str) -> AlertStrategy:
    return STRATEGIES.get(component_type.upper(), AlertStrategy())

File: backend/src/domain/alertStrategies.js
export class AlertStrategy {
  evaluate() {
    return { severity: "P3", channel: "dashboard", responder: "sre-oncall" };
  }
}

export class RdbmsAlertStrategy extends AlertStrategy {
  evaluate() {
    return { severity: "P0", channel: "pager", responder: "database-oncall" };
  }
}

export class McpAlertStrategy extends AlertStrategy {
  evaluate() {
    return { severity: "P1", channel: "pager", responder: "platform-oncall" };
  }
}

export class CacheAlertStrategy extends AlertStrategy {
  evaluate() {
    return { severity: "P2", channel: "slack", responder: "cache-oncall" };
  }
}

export class QueueAlertStrategy extends AlertStrategy {
  evaluate() {
    return { severity: "P1", channel: "pager", responder: "async-oncall" };
  }
}

const strategies = {
  RDBMS: new RdbmsAlertStrategy(),
  MCP_HOST: new McpAlertStrategy(),
  CACHE: new CacheAlertStrategy(),
  QUEUE: new QueueAlertStrategy()
};

export function getAlertStrategy(componentType = "") {
  return strategies[componentType.toUpperCase()] ?? new AlertStrategy();
}
